ignite/metrics/mAP.py

```diff
@@ -0,0 +1,402 @@
+
+                for t, (tp, fp) in enumerate(zip(tp_sum, fp_sum)):
+                    nd = len(tp)
+                    rc = tp / non_ignored
+                    pr = tp / (fp + tp + torch.finfo(torch.float64).eps)
+                    q: Any = torch.zeros((num_rec_thr,))
+
+                    pr = pr.tolist()
+                    q = q.tolist()
+
+                    for i in range(nd - 1, 0, -1):
+                        if pr[i] > pr[i - 1]:
+                            pr[i - 1] = pr[i]
+
+                    inds = torch.searchsorted(rc, self.rec_thrs, right=False)
+                    try:
+                        for ri, pi in enumerate(inds):
+                            q[ri] = pr[pi]
+                    except:
+                        pass
+                    precision[t, :, c, a] = torch.tensor(q)
+        # return precision, recall
+        self.precision = precision
+
+    def _summarize(self, iou_thr: Optional[float] = None, area_rng: str = "all") -> float:
+        aind = [i for i, a_rng in enumerate(self.area_rngs) if a_rng == area_rng]
+        # Calculate Average Precision
+        s = self.precision
+        if iou_thr is not None:
+            t = (self.iou_thrs == iou_thr).int().nonzero(as_tuple=True)[0]
+            s = s[t]
+        s = s[:, :, :, aind]
+        # Take mean to calculate mAP
+        if len(s[s > -1]) == 0:
+            mean_s = torch.tensor(-1)
+        else:
+            mean_s = torch.mean(s[s > -1])
+
+        return mean_s.item()
+
+    def _summarize_all(self) -> torch.Tensor:
+        stats = torch.zeros((6,))
+        stats[0] = self._summarize()
+        stats[1] = self._summarize(iou_thr=0.5)
+        stats[2] = self._summarize(iou_thr=0.75)
+        stats[3] = self._summarize(area_rng="small")
+        stats[4] = self._summarize(area_rng="medium")
+        stats[5] = self._summarize(area_rng="large")
+
+        return stats
+
+    @reinit__is_reduced
+    def _gather_all(self) -> None:
+        import torch.distributed as dist
+
+        def is_dist_avail_and_initialized() -> bool:
+            if not dist.is_available():
+                return False
+            if not dist.is_initialized():
+                return False
+            return True
+
+        if not is_dist_avail_and_initialized():
+            return
+
+        gather_dicts: List[Dict] = [defaultdict(list)] * dist.get_world_size()
+        dist.gather_object(self.eval_imgs, gather_dicts if dist.get_rank() == 0 else None, dst=0)
+        if dist.get_rank() == 0:
+            keys = set()
+            for eval_imgs in gather_dicts:
+                for key in eval_imgs:
+                    keys.add(key)
+            combined_eval_imgs: defaultdict = defaultdict(list)
+            for key in keys:
+                for eval_imgs in gather_dicts:
+                    combined_eval_imgs[key] += eval_imgs[key]
+            self.eval_imgs = combined_eval_imgs
+
+    def compute(self) -> torch.Tensor:
```

Comment:
Could the output be a dict rather than a tensor ? It would help the user and ease the print of the results